ttnn/cpp/ttnn/kernel/kernel_utils.hpp

```diff
@@ -0,0 +1,40 @@
+// SPDX-FileCopyrightText: © 2025 Tenstorrent Inc.
+//
+// SPDX-License-Identifier: Apache-2.0
+#pragma once
+
+#if defined(KERNEL_BUILD)
+
+#if defined(COMPILE_FOR_NCRISC) || defined(COMPILE_FOR_BRISC)
+#include "dataflow_api.h"
+#elif defined(COMPILE_FOR_TRISC)
+#include "compute_kernel_api/common.h"
+#endif
+
+#include <utility>
+#include "compile_time_args.h"
+
+namespace ttnn::kernel_utils {
+template <typename KernelArgsStruct, uint32_t... I>
+KernelArgsStruct make_runtime_struct_from_args(std::integer_sequence<uint32_t, I...>) {
+    return KernelArgsStruct{get_arg_val<uint32_t>(I)...};
+}
+
+template <typename KernelArgsStruct>
+KernelArgsStruct make_runtime_struct_from_args() {
+    constexpr uint32_t num_fields = sizeof(KernelArgsStruct) / sizeof(uint32_t);
+    return make_runtime_struct_from_args<KernelArgsStruct>(std::make_integer_sequence<uint32_t, num_fields>{});
+}
+
+template <typename KernelArgsStruct, uint32_t... I>
+constexpr KernelArgsStruct make_compile_time_struct_from_args(std::integer_sequence<uint32_t, I...>) {
+    return KernelArgsStruct{get_compile_time_arg_val(I)...};
+}
+
+template <typename KernelArgsStruct>
+constexpr KernelArgsStruct make_compile_time_struct_from_args() {
+    constexpr uint32_t num_fields = sizeof(KernelArgsStruct) / sizeof(uint32_t);
+    return make_compile_time_struct_from_args<KernelArgsStruct>(std::make_integer_sequence<uint32_t, num_fields>{});
+}
```

Comment:
This logic is wrong. Just because the concept defined earlier is satisfied does not mean the number of fields is the same as the number of `uint32_t`, so you can't use aggregate initialization.

```suggestion
namespace ttnn::kernel_utils {
template <typename T, uint32_t... I>
T make_runtime_struct_from_args(std::integer_sequence<uint32_t, I...>) {
    const uint32_t args[]{get_arg_val<uint32_t>(I)...};
    return reinterpret_cast<const T &>(args);
}

template <typename T>
T make_runtime_struct_from_args() {
    constexpr uint32_t num_fields = sizeof(T) / sizeof(uint32_t);
    return make_runtime_struct_from_args<KernelArgsStruct>(std::make_integer_sequence<uint32_t, num_fields>{});
}

template <typename T, uint32_t... I>
T make_compile_time_struct_from_args(std::integer_sequence<uint32_t, I...>) {
    constexpr uint32_t args[]{get_compile_time_arg_val(I)...};
    return reinterpret_cast<const T &>(args);
}

template <typename T>
T make_compile_time_struct_from_args() {
    constexpr uint32_t num_fields = sizeof(T) / sizeof(uint32_t);
    return make_compile_time_struct_from_args<T>(std::make_integer_sequence<uint32_t, num_fields>{});
}
```

Reply:
I previously checked that all fields were uint32_t using reflection, but I’ve removed that check. Looking at it now, we don’t need that restriction at all – I like this approach, thank you for the suggestion.

Using reinterpret_cast results in UB, which also prevents calling the function in a constexpr context. I can’t use std::bit_cast in the kernel (error: 'bit_cast' is not a member of 'std'), but it seems __builtin_bit_cast works.

What do you think about __builtin_bit_cast usage ?